### PR TITLE
e2e pods: fix WaitForPodsResponding retry

### LINKED ISSUE #120539: e2e WaitForPodsResponding does not retry
Seen on the CI here https://github.com/cilium/cilium/issues/25655 in this specific job https://github.com/cilium/cilium/issues/25655#issuecomment-1712214678

The apiserver returns 503 

```
[FAILED] checking pod responses: Told to stop trying after 0.155s.
Unexpected final error while getting []pod.response: Controller slow-terminating-unready-pod: failed to Get from replica pod slow-terminating-unready-pod-x78jm:
    <*errors.StatusError | 0xc003a72500>: 
    the server is currently unable to handle the request (get pods slow-terminating-unready-pod-x78jm)
    {
        ErrStatus: 
            code: 503
            details:
              causes:
              - message: unknown
                reason: UnexpectedServerResponse
              kind: pods
              name: slow-terminating-unready-pod-x78jm
            message: the server is currently unable to handle the request (get pods slow-terminating-unready-pod-x78jm)
            metadata: {}
            reason: ServiceUnavailable
            status: Failure,
    }
Pod status:
    <v1.PodStatus>: 
        conditions:
        - lastProbeTime: null
          lastTransitionTime: "2023-09-08T20:07:05Z"
          status: "True"
          type: PodScheduled
        phase: Pending
        qosClass: BestEffort
In [It] at: test/e2e/network/service.go:1816 @ 09/08/23 20:07:33.779
```

and the [function fails inmediatly] (https://pipelinesghubeus22.actions.githubusercontent.com/0fSQMQEEZURuRkH6hfAJzqwr7ktdlqLQsBeuJ3PrFLFXXTDS40/_apis/pipelines/1/runs/1812127/signedlogcontent/2?urlExpires=2023-09-09T08%3A00%3A16.5542540Z&urlSigningMethod=HMACV1&urlSignature=TA7opAGGyxgZGQ3vojFGazyNbf7K8bdeL0q20cSfcTM%3D) instead of retrying

```
2023-09-08T20:07:39.5813786Z   Sep  8 20:07:05.701: INFO: Pod name slow-terminating-unready-pod: Found 1 pods out of 1
2023-09-08T20:07:39.5817086Z   [1mSTEP:[0m ensuring each pod is running [38;5;243m@ 09/08/23 20:07:05.701[0m
2023-09-08T20:07:39.5823170Z   [1mSTEP:[0m trying to dial each unique pod [38;5;243m@ 09/08/23 20:07:33.163[0m
2023-09-08T20:07:39.5826432Z   Sep  8 20:07:33.319: INFO: Failed inside E2E framework:
2023-09-08T20:07:39.5830316Z       k8s.io/kubernetes/test/e2e/framework/pod.WaitForPodsResponding({0x7f3f683419b0, 0xc003c399e0}, {0x72b8af0?, 0xc001529d40}, {0xc001e45bf0, 0xd}, {0x6b1917f, 0x1c}, 0x0, 0xd18c2e2800, ...)
2023-09-08T20:07:39.5836285Z       	test/e2e/framework/pod/wait.go:625 +0x3ef
2023-09-08T20:07:39.5837043Z       k8s.io/kubernetes/test/e2e/framework/pod.podRunningMaybeResponding({0x7f3f683419b0, 0xc003c399e0}, {0x72b8af0, 0xc001529d40}, {0xc001e45bf0, 0xd}, {0x6b1917f, 0x1c}, 0x0?, 0x1, ...)
2023-09-08T20:07:39.5841583Z       	test/e2e/framework/pod/resource.go:117 +0x11d
2023-09-08T20:07:39.5842322Z       k8s.io/kubernetes/test/e2e/framework/pod.VerifyPods(...)
2023-09-08T20:07:39.5845500Z       	test/e2e/framework/pod/resource.go:99
2023-09-08T20:07:39.5846007Z       k8s.io/kubernetes/test/e2e/network.glob..func27.22({0x7f3f683419b0, 0xc003c399e0})
2023-09-08T20:07:39.5849411Z       	test/e2e/network/service.go:1816 +0xbfb

```

https://github.com/kubernetes/kubernetes/blob/21f7bf66fa949dda2b3bec6e3581e248e270e001/test/e2e/framework/pod/wait.go#L621-L626

I expected my fix https://github.com/kubernetes/kubernetes/commit/7888798873e1b87e320b8714e17ef67af7c25ec0 to retry on those errors, but is clearly not the case

@onsi @pohly I find very hard to understand the logic to retry in this function, can you PTAL and identify the reason why that function is not retrying on these errors?

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:

The status error was embedded inside the new error constructed by WaitForPodsResponding's get function, but not wrapped. Therefore `apierrors.IsServiceUnavailable(err)` didn't find it and returned false -> no retries.

Wrapping fixes this and Gomega formatting of the error remains useful:

	err := &errors.StatusError{}
	err.ErrStatus.Code = 503
	err.ErrStatus.Message = "temporary failure"

	err2 := fmt.Errorf("Controller %s: failed to Get from replica pod %s:\n%w\nPod status:\n%s",
		"foo", "bar",
		err, "some status")
	fmt.Println(format.Object(err2, 1))
        fmt.Println(errors.IsServiceUnavailable(err2))

=>

    <*fmt.wrapError | 0xc000139340>:
    Controller foo: failed to Get from replica pod bar:
    temporary failure
    Pod status:
    some status
    {
        msg: "Controller foo: failed to Get from replica pod bar:\ntemporary failure\nPod status:\nsome status",
        err: <*errors.StatusError | 0xc0001a01e0>{
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "",
                Message: "temporary failure",
                Reason: "",
                Details: nil,
                Code: 503,
            },
        },
    }

    true

#### Which issue(s) this PR fixes:
Fixes #120539

#### Does this PR introduce a user-facing change?
```release-note
e2e framework: retrying after intermittent apiserver failures was fixed in WaitForPodsResponding
```

/cc @alculquicondor

### CODE REVIEW

File: test/e2e/framework/pod/wait.go [hunk 1]
```diff
@@ -581,10 +581,11 @@ func WaitForPodsResponding(ctx context.Context, c clientset.Interface, ns string
 
 			if err != nil {
 				// We may encounter errors here because of a race between the pod readiness and apiserver
-				// proxy. So, we log the error and retry if this occurs.
-				return nil, fmt.Errorf("Controller %s: failed to Get from replica pod %s:\n%s\nPod status:\n%s",
+				// proxy or because of temporary failures. The error gets wrapped for framework.HandleRetry.
+				// Gomega+Ginkgo will handle logging.
+				return nil, fmt.Errorf("Controller %s: failed to Get from replica pod %s:\n%w\nPod status:\n%s",
 					controllerName, pod.Name,
-					format.Object(err, 1), format.Object(pod.Status, 1))
```
Comment (aojea) [inline test/e2e/framework/pod/wait.go:587]: oh, very subtle, impressive

**Comment by aojea** [comment]
/lgtm

Thanks, can we backport it?

**Comment by aojea** [comment]
linter complain

> ERROR: test/e2e/framework/pod/wait.go:586:17: ST1005: error strings should not be capitalized (stylecheck)
ERROR: 				return nil, fmt.Errorf("Controller %s: failed to Get from replica pod %s:\n%w\nPod status:\n%s",

**Comment by pohly** [comment]
> linter complain

Fixed.

> Thanks, can we backport it?

I agree, we should. Let's merge it and I'll prepare cherry-picks.

**Comment by pohly** [comment]
Backports:
* 1.28: https://github.com/kubernetes/kubernetes/pull/120572
* 1.27: https://github.com/kubernetes/kubernetes/pull/120573
* 1.26 and older: not applicable (was still using wait.Poll)

**Pull request merged**